src/A2A/Models/A2AResponse.cs

```diff
@@ -71,61 +71,18 @@ public abstract class A2AEvent(A2AEventKind kind)
 // we implement our own converter to handle that, with the discriminator logic implemented by-hand.
 public abstract class A2AResponse(A2AEventKind kind) : A2AEvent(kind);
 
-internal class A2AEventConverterViaKindDiscriminator<T> : JsonConverter<T> where T : A2AEvent
+internal class A2AEventConverterViaKindDiscriminator<T> : BaseKindDiscriminatorConverter<T, A2AEventKind> where T : A2AEvent
 {
-    internal const string DiscriminatorPropertyName = "kind";
-
-    public override T Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
-    {
-        using var document = JsonDocument.ParseValue(ref reader);
-        var root = document.RootElement;
-
-        if (!root.TryGetProperty(DiscriminatorPropertyName, out var kindProp) || kindProp.ValueKind is not JsonValueKind.String)
-        {
-            throw new A2AException($"Missing required '{DiscriminatorPropertyName}' discriminator for {typeof(T).Name}.", A2AErrorCode.InvalidRequest);
-        }
-
-        T? a2aEventObj = null;
-        Exception? deserializationException = null;
-        try
-        {
-            var kindValue = kindProp.Deserialize(A2AJsonUtilities.JsonContext.Default.A2AEventKind);
-#pragma warning disable CS8524 // The switch expression does not handle some values of its input type (it is not exhaustive) involving an unnamed enum value.
-            // We don't need to handle this because the previous Deserialize call would have thrown if the value was invalid.
-            JsonTypeInfo typeInfo = kindValue switch
-            {
-                A2AEventKind.Message => options.GetTypeInfo(typeof(AgentMessage)),
-                A2AEventKind.Task => options.GetTypeInfo(typeof(AgentTask)),
-                A2AEventKind.StatusUpdate => options.GetTypeInfo(typeof(TaskStatusUpdateEvent)),
-                A2AEventKind.ArtifactUpdate => options.GetTypeInfo(typeof(TaskArtifactUpdateEvent)),
-            };
-#pragma warning restore CS8524 // The switch expression does not handle some values of its input type (it is not exhaustive) involving an unnamed enum value.
-
-            a2aEventObj = (T?)root.Deserialize(typeInfo);
-        }
-        catch (Exception e)
-        {
-            deserializationException = e;
-        }
-
-        if (deserializationException is not null || a2aEventObj is null)
-        {
-            throw new A2AException($"Failed to deserialize {kindProp.GetString()} event", deserializationException, A2AErrorCode.InvalidRequest);
-        }
-
-        return a2aEventObj;
-    }
-
-    public override void Write(Utf8JsonWriter writer, T value, JsonSerializerOptions options)
+    protected override Type?[] TypeMapping => new Type?[]
```

Comment:
This allocates with every invocation. Please consider to create map once as static or some other implementation. For example instead of  `protected abstract Type?[] TypeMapping { get; }` handle it by `GetTypeFromKind(TKind)` which could be implemented by simple switch as we do not expect lots of derived classes. One way or another please try to avoid unnecessary allocations per deser object instance.